Tribler/community/tunnel/processes/processmanager.py

```diff
@@ -0,0 +1,186 @@
+import logging
+import os
+import shutil
+from multiprocessing import cpu_count
+from random import sample
+
+from Tribler.community.tunnel.processes.tunnel_childprocess import TunnelProcess
+
+
+class ProcessManager(object):
+
+    """The ProcessManager creates and manages ChildProcesses
+    """
+
+    def __init__(self, session, community=None):
+        """Create a new ProcessManager
+
+        :param session: the session to extract info from
+        :type session: Tribler.Core.Session.Session
+        :param community: (non-Session) community to report to
+        :type community: Tribler.dispersy.community.Community
+        :returns: None
+        """
+        super(ProcessManager, self).__init__()
+
+        self.pool = {}          # Map of pid -> Process
+        self.circuit_map = {}   # Map of circuit -> pid
+        self.session = session
+        self.community = community
+
+        self._clean_working_dir()
+
+    def _clean_working_dir(self):
+        """Clean leftover directories from crashed/terminated processes
+
+        :returns: None
+        """
+        for directory in [x[0] for x in os.walk(self.session.get_state_dir())]:
+            if os.path.split(directory)[1].startswith("SUBPROCESS"):
+                logging.error("Cleaning up leftover subprocess artifacts in " + directory)
+                try:
+                    shutil.rmtree(directory)
+                except OSError:
+                    pass
+
+    def get_suggested_workers(self):
+        """Have the process manager suggest a number of workers to use
+
+        :return: the suggested amount of workers
+        :rtype: int
+        """
+        return cpu_count()
+
+    def set_worker_count(self, value):
+        """Set the amount of workers to use
+
+        :param value: the new amount of workers
+        :returns: None
+        """
+        count = self.get_worker_count()
+        if count < value:
+            # We have too little workers, create more
+            for _ in range(value - count):
+                def on_worker(worker):
+                    self.pool[worker.pid] = worker
+                self._create_worker(on_worker)
+        elif count > value:
+            # We have too many workers, remove some
+            self._remove_workers(max(count - value, count))
+
+    def get_worker_count(self):
+        """Return the active amount of workers
+
+        :return: the current worker count
+        :rtype: int
+        """
+        return len(self.pool.keys())
+
+    def _create_worker(self, callback):
+        """Create a single worker, without adding it to the pool
```

Comment:
Otherwise external classes might select a worker from the pool which has not properly been initialized yet.